core/src/main/java/org/mskcc/cbio/portal/servlet/PathologyReportView.java

```diff
@@ -0,0 +1,224 @@
+            validateRequest(request);
+            if (request.getAttribute(ERROR_CODE)!=null) {
+                response.sendError(Integer.parseInt((String) request.getAttribute(ERROR_CODE)), 
+                    (String) request.getAttribute(ERROR_MSG));
+            } else {
+                File pathologyReport = getRequestedFile(request);
+                serveFile(response, pathologyReport);
+            }
+        } catch (DaoException e) {
+            logger.error("Got Database Exception while processing request for pathology report.", e);
+            response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, 
+                "An error occurred while trying to connect to the database.");
+        } catch (NoSuchFileException e) {
+            logger.error("Got No Such File Exception while processing request for pathology report.", e);
+            response.sendError(HttpServletResponse.SC_NOT_FOUND, 
+                "The requested pathology report could not be located.");
+        }
+    }
+    
+    private String getRequestedPath(HttpServletRequest request) throws IOException{
+        return URLDecoder.decode(request.getPathInfo().substring(1), "UTF-8");
+    }
+    
+    private File getRequestedFile(HttpServletRequest request) throws IOException {
+        // Convert from the expected url path 'studyId/patientId/sampleId.pdf' to the expected filesystem path 'studyId.patientId.sampleId.pdf'
+        final String convertedPath = getRequestedPath(request).replace("/", ".");
+        return new File(DATA_DIRECTORY, convertedPath);
+    }
+    
+    private void setError(HttpServletRequest request, int httpStatusCode, String errorMessage) {
+        request.setAttribute(ERROR_CODE, Integer.toString(httpStatusCode));
+        request.setAttribute(ERROR_MSG, errorMessage);
+    }
+    
+    private boolean validateRequest(HttpServletRequest request) throws IOException, DaoException {
+        final String requestedPath = getRequestedPath(request);
+        final String[] path = requestedPath.split("/");
+        if (path.length != 3) {
+            setError(request, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Unable to process requested pathology " +
+                "report path. Ensure it is in the format cbioportal/pathology_report/study_id/patient_id/sample_id.pdf");
+            return false;
+        }
+
+        final String cancerStudyId = path[0];
+        CancerStudy cancerStudy = DaoCancerStudy.getCancerStudyByStableId(cancerStudyId);
+        if (cancerStudy == null) {
+            setError(request, HttpServletResponse.SC_NOT_FOUND, "No such cancer study with id '" + cancerStudyId + "'.");
+            return false;
+        }
+        if (accessControl.isAccessibleCancerStudy(cancerStudyId).size() != 1) {
+            setError(request, HttpServletResponse.SC_FORBIDDEN, 
+                "You are not authorized to access the cancer study with id '" + cancerStudyId + "'.");
+            return false;
```

Comment:
The multiple return statements are intended for exiting the method early when an error is found and written to a request variable. An alternative may be to define some kind of new validation exception that has an error code and message, throw it and then handle it within the process request function (rather than use the request attribute for recording and accessing the error). Do you think this would be a better approach or stick with what is there already?